File: chap1/Cpp/03_gaxpy.cpp
#include <armadillo>
#include <cassert>

using namespace arma;

void gaxpy(const mat &A, const vec &x, vec &y) {
    auto m = A.n_rows;
    auto n = A.n_cols;
    assert(n == x.size() && m == y.size());
    
    for (int i=0; i<m; i++) {
        for (int j=0; j<n; j++) {
            y[i] += A(i,j) * x[j];
        }
    }
}

int main() {
    mat A(2, 2);
    A << 1 << 2 << endr
        << 3 << 4 << endr;
    vec x = {1, 2};
    vec y = {0, 0};
    gaxpy(A, x, y);
    y.print();
}

File: chap1/Rust/src/bin/03_gaxpy.rs
#[macro_use]
extern crate peroxide;
use peroxide::fuga::*;

#[allow(non_snake_case)]
fn main() {
    let A = ml_matrix("1 2;3 4");
    let x = c!(1,2);
    let mut y = c!(0,0);
    gaxpy(&A, &x, &mut y);
    y.print();
}

#[allow(non_snake_case)]
fn gaxpy(A: &Matrix, x: &Vec<f64>, y: &mut Vec<f64>) {
    let m = A.row;
    let n = A.col;
    assert!(m == y.len() && n == x.len());

    for i in 0 .. m {
        for j in 0 .. n {
            y[i] += A[(i, j)] * x[j];
        }
    }
}
